### PR TITLE
Sync Type(Struct|Class).dotExp

### PR DESCRIPTION
**Pull request by 9rnsr**

### CODE REVIEW

File: src/mtype.d [hunk 1]
```diff
@@ -7803,14 +7803,11 @@ public:
...
         }
+
         // Bugzilla 14010
         if (ident == Id._mangleof)
             return getProperty(e.loc, ident, flag);
-        if (!sym.members)
-        {
-            error(e.loc, "struct %s is forward referenced", sym.toChars());
-            return new ErrorExp();
-        }
```
Comment (9rnsr) [inline src/mtype.d]: Whether an opaque struct, is already checked in `AggregateDeclaration.size` (--> `determineSize`) and `StructDeclaration.search`.

File: src/mtype.d [hunk 2]
```diff
@@ -7819,7 +7816,8 @@ public:
              * (e.field0, e.field1, e.field2, ...)
              */
             e = e.semantic(sc); // do this before turning on noaccesscheck
-            e.type.size(); // do semantic of type
+
+            sym.size(e.loc); // do semantic of type
```
Comment (9rnsr) [inline src/mtype.d]: `TypeStruct.size()` size calls `sym.size`.

File: src/mtype.d [hunk 3]
```diff
@@ -7849,30 +7847,12 @@ public:
...
         if (!s)
         {
-            if (sym._scope) // it's a fwd ref, maybe we can resolve it
-            {
-                sym.semantic(null);
-                s = sym.search(e.loc, ident);
-            }
-            if (!s)
-                return noMember(sc, e, ident, flag);
+            return noMember(sc, e, ident, flag);
```
Comment (9rnsr) [inline src/mtype.d]: `sym.semantic` is already called in `sym.search`.

File: src/mtype.d [hunk 4]
```diff
@@ -7963,13 +7948,14 @@ public:
...
             return new DotExp(e.loc, e, oe);
         }
+
         Declaration d = s.isDeclaration();
-        debug
+        if (!d)
         {
-            if (!d)
-                printf("d = %s '%s'\n", s.kind(), s.toChars());
+            e.error("%s.%s is not a declaration", e.toChars(), ident.toChars());
```
Comment (WalterBright) [inline src/mtype.d]: The reason for the `assert(d)` was that this should never happen. Replacing it with `error` means it does happen - but there is no test case showing it can be.

File: src/mtype.d [hunk 5]
```diff
@@ -8564,39 +8551,24 @@ public:
...
-        {
-            DotExp de = cast(DotExp)e;
-            if (de.e1.op == TOKscope)
-            {
-                ScopeExp se = cast(ScopeExp)de.e1;
-                s = se.sds.search(e.loc, ident);
-                e = de.e1;
-                goto L1;
-            }
+            printf("TypeClass::dotExp(e = '%s', ident = '%s')\n", e.toChars(), ident.toChars());
```
Comment (WalterBright) [inline src/mtype.d]: Adding spaces is pointless, and causes problems with line wrap when the expressions are long.

File: src/mtype.d [hunk 6]
```diff
@@ -8564,39 +8551,24 @@ public:
         Dsymbol s;
         static if (LOGDOTEXP)
         {
-            printf("TypeClass::dotExp(e='%s', ident='%s')\n", e.toChars(), ident.toChars());
-        }
-        if (e.op == TOKdot)
```
Comment (WalterBright) [inline src/mtype.d]: why is this removed

File: src/mtype.d [hunk 7]
```diff
@@ -8850,21 +8832,19 @@ public:
...
             e.error("%s.%s is not a declaration", e.toChars(), ident.toChars());
             return new ErrorExp();
         }
+
         if (e.op == TOKtype)
         {
             /* It's:
              *    Class.d
              */
-            // If Class is in a failed template, return an error
```
Comment (WalterBright) [inline src/mtype.d]: why is this removed

File: src/mtype.d [hunk 8]
```diff
@@ -8952,14 +8934,6 @@ public:
             e = e.semantic(sc);
             return e;
         }
-        if (d.parent && d.toParent().isModule())
```
Comment (WalterBright) [inline src/mtype.d]: why is this removed?

File: src/mtype.d (hunk 2 above)
Comment (WalterBright) [inline src/mtype.d]: maybe `e.type` isn't a struct

File: src/mtype.d [hunk 9]
```diff
@@ -7849,30 +7847,12 @@ public:
             sc2.pop();
             return e;
         }
-        if (e.op == TOKdot)
```
Comment (WalterBright) [inline src/mtype.d]: why is this removed

File: src/mtype.d (hunk 3 above)
Comment (WalterBright) [inline src/mtype.d]: Then replace the code with an `assert` that it is not forward referenced.

**Comment by WalterBright** [comment]
This is not a refactor, as there are many semantic changes.

File: src/mtype.d (hunk 2 above)
Comment (9rnsr) [inline src/mtype.d]: The necessary thing to do here is, to call `StructDeclaration.size`. Even if `e.type` is Tstruct (e.g. `e.type.ty == Tenum && e.type.tobasetype().ty == Tstruct`?), we don't have to call `e.type.size` in here.

File: src/mtype.d (hunk 4 above)
Comment (9rnsr) [inline src/mtype.d]: I understand, but the same case is handled by `error` in `TypeClass.dotExp`. I couldn't find a related test case for the error report in `TypeClass.dotExp`, so I concluded that this line would have possibility to cause assertion failure in released compiler, because whena symbol can be declared in class members, equivalent symbol can also be declared in struct member.

Honestly, similar error report exists at the end of `DsymbolExp.resolve` function (`"%s '%s' is not a variable"`). In old days I thought it would be unreachable and could be replaced with `assert`, but it was actually reached because of other place bug (see [issue 14858](https://issues.dlang.org/show_bug.cgi?id=14858)). I think this replacement is better so it's more defensive.

File: src/mtype.d (hunk 6 above)
Comment (9rnsr) [inline src/mtype.d]: `DotExp` is already handled in `DotIdExp.semanticY`, so this code is unreachable today.

File: src/mtype.d (hunk 5 above)
Comment (9rnsr) [inline src/mtype.d]: When I comapre the two functions code by using diff tool, here had had a small difference. Adding spaces around operator `=` is normal code style in dmd code base, so I synced the debug message with:
```d
printf("TypeStruct::dotExp(e = '%s', ident = '%s')\n", e.toChars(), ident.toChars());
```

File: src/mtype.d (hunk 7 above)
Comment (9rnsr) [inline src/mtype.d]: Because this is not follow current error handling way.

Although a class is instantiated and its parent has errors, when the class member `d` does not directly depend on that parent errors, the type of `d` can be valid, and useful error message will be shown. For example:

```d
template C(T)  // tiparent == C!T
{
    /* some error occurences make TemplateInstance('C!T().errors true. */

    class C
    {
        int d;

        void test()
        {
            auto c = new C();
            c.d = "abc";  // Error: string "abc" is not convertible to int

            // In TypeClass.dotExp, d == VarDeclaration('C!T.C.d')
        }
    }
}
```

The incompatible type error is not related with the errors in `tiparent`, so showing two unrelated errors is better diagnostic.

File: src/mtype.d (hunk 8 above)
Comment (9rnsr) [inline src/mtype.d]: If `d` is VarDeclaration, `d.isDataseg()` would return true, and it's already handled in previous lines.
If `d` is FuncDeclaration, we have to consider it would have overloads. For example:

```d
module x;
void foo() {}

class C
{
    alias bar = .foo;
    void bar(int) {}
}

void main()
{
    auto c = new C();
    c.bar(1);
}
```

By removing this, such the dificult case handlings will be gathered into `DotVarExp.semantic`.

**Comment by 9rnsr** [comment]
> This is not a refactor, as there are many semantic changes.

OK, I removed the word "refactor" from the summary.

File: src/mtype.d (hunk 3 above)
Comment (9rnsr) [inline src/mtype.d]: I hope to completely encapsule the forward reference handling into `search` member function, so I don't want to add such the redundant assert in here.

File: src/mtype.d (hunk 9 above)
Comment (9rnsr) [inline src/mtype.d]: Same reason for `TypeClass.dotExp`.

File: src/mtype.d (hunk 1 above)
Comment (WalterBright) [inline src/mtype.d]: Should add an assert, then. Asserts are cheap, they document assumptions, and check them. dmd is very complex already, and so this is doubly important.

File: src/mtype.d (hunk 3 above)
Comment (WalterBright) [inline src/mtype.d]: 1. Asserts are cheap
2. Assets are by definition redundant
3. Future work is not in this PR

File: src/mtype.d (hunk 4 above)
Comment (WalterBright) [inline src/mtype.d]: I'd prefer to leave it as an assert until someone finds a case that trips it.

File: src/mtype.d (hunk 6 above)
Comment (WalterBright) [inline src/mtype.d]: add an assert then to ensure it.

File: src/mtype.d (hunk 8 above)
Comment (WalterBright) [inline src/mtype.d]: > If d is VarDeclaration, d.isDataseg() would return true

This is not correct. isDataseg() returns true for only some VarDeclaration's.

File: src/mtype.d (hunk 1 above)
Comment (9rnsr) [inline src/mtype.d]: This is not able to replace with assert, because in here, `sym,members` could be `null`.

File: src/mtype.d (hunk 3 above)
Comment (9rnsr) [inline src/mtype.d]: This is not future work. It's already proven in other places where the `StructDeclaration.search` is called only once. The double `search` call was added in very old ages, and it's not necessary today. It's a time to remove the old age garbage.

File: src/mtype.d (hunk 4 above)
Comment (9rnsr) [inline src/mtype.d]: My concern is: different from c-dmd, assert is useless with **released** d-dmd. Currently it's compiled with -release, so assert is never tested. Therefore if `d` would become `null`, it would cause OS-specific segmentation fault. The added error message is definitely better than that.

File: src/mtype.d (hunk 8 above)
Comment (9rnsr) [inline src/mtype.d]: > isDataseg() returns true for only some VarDeclaration's.

I know it, but the other cases - the `e` and `d` needs to be combined with `CommaExp` - should be handled in `DotVarExp`. If its current code is not sufficient, it would be future work.

File: src/mtype.d (hunk 4 above)
Comment (9rnsr) [inline src/mtype.d]: One idea to emphasize it's not expected to reach the error, is use special error message "Internal Compiler Error: ...". How about that?

File: src/mtype.d (hunk 6 above)
Comment (9rnsr) [inline src/mtype.d]: I added `assert(e.op != TOKdot);` both in `Type(Struct|Class).dotExp`.

File: src/mtype.d (hunk 4 above)
Comment (yebblies) [inline src/mtype.d]: Are we really shipping with `-release`?  That's not idea.

Comment (9rnsr) [inline src/mtype.d]: In `src/win32.mak`:

```
dmd: reldmd

release:
	$(DMDMAKE) clean
	$(DMDMAKE) reldmd
	$(DMDMAKE) clean

debdmd:
	$(DMDMAKE) "OPT=" "DEBUG=-D -g -DUNITTEST" "DDEBUG=-debug -g -unittest" "DOPT=" "LFLAGS=-L/ma/co/la" $(TARGETEXE)

reldmd:
	$(DMDMAKE) "OPT=-o" "DEBUG=" "DDEBUG=" "DOPT=-O -release -inline" "LFLAGS=-L/delexe/la" $(TARGETEXE)
```

`reldmd` is built with `-release`.

**Comment by WalterBright** [comment]
Auto-merge toggled on
